### PR TITLE
Treat task inputs as arguments

### LINKED ISSUE #288: Task arguments act like inputs
for example they perform the required logic and in cast that a required input is not available fail the flow instead of passing the value as-is

### PR DESCRIPTION
**Pull request by Bonczidai**
fixes #288 

Signed-off-by: Bonczidai Levente <levente.bonczidai@hp.com>

### CODE REVIEW

**Comment by meirwah** [comment]
@orius123 please take a look (even that it's not in ready to merge)

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/scorecompiler/ScoreCompilerImpl.java [hunk 1]
```diff
@@ -142,11 +142,12 @@ private ExecutionPlan compileToExecutionPlan(Executable executable) {
 		Collection<Input> result = new ArrayList<>();
 		for(Executable executable : executables) {
 			result.addAll(getSystemProperties(executable.getInputs()));
-			if(executable instanceof Flow) {
-				for(Task task : ((Flow)executable).getWorkflow().getTasks()) {
-					result.addAll(getSystemProperties(task.getInputs()));
-				}
-			}
+            // TODO - task args - don't support sys prop for task args
```
Comment (orius123) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/scorecompiler/ScoreCompilerImpl.java]: why comment out? you can just delete it..

Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/scorecompiler/ScoreCompilerImpl.java]: correct - I am pushing now some changes regarding the compiler (enhancements, test updates)

Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/scorecompiler/ScoreCompilerImpl.java]: done

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/transformers/DoTransformer.java [hunk 2]
```diff
@@ -58,4 +54,33 @@ public String keyToTransform() {
...
+    private Argument transformArgument(Object rawArgument) {
+        // - some_arg
+        // this is our default behaviour that if the user specifies only a key, the key is also the ref we look for
+        if (rawArgument instanceof String) {
+            String argumentName = (String) rawArgument;
+            return new Argument(argumentName, null);
+        } else if (rawArgument instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map.Entry<String, ?> entry = ((Map<String, ?>) rawArgument).entrySet().iterator().next();
+            Object entryValue = entry.getValue();
```
Comment (orius123) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/transformers/DoTransformer.java]: make it Serializable

**Comment by Bonczidai** [comment]
some notes on one liner `#2` syntax:

```yaml
containers.get_filtered_containers: docker_options, all_containers = true, excluded_images = 'STR that contains \, comma'
```

task arguments are separated via `,`, name and expression is separated via `=`
`,` can be used in expression by `\,` escaping it, expression will be evaluated *as is* (no double quotes for strings) -- see test cases

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/transformers/DoTransformer.java [hunk 3]
```diff
@@ -58,4 +61,64 @@ public String keyToTransform() {
...
+        }
+        throw new RuntimeException("Could not transform task argument: " + rawArgument);
+    }
+
+    private List<Argument> transformOneLinerArguments(String line) {
+        List<Argument> arguments = new ArrayList<>();
+
+        List<String> rawArguments = Arrays.asList(
+                // every comma that is not preceded by '\' (escaped)
+                line.split("(?<!\\\\)(,)")
```
Comment (meirwah) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/transformers/DoTransformer.java]: take regex "(?<!\\\\)(,)"  to external const string param...

File: cloudslang-compiler/src/test/resources/flow_implicit_alias_for_current_namespace.sl [hunk 4]
```diff
@@ -17,12 +17,10 @@ flow:
         do:
           ops.test_op:
             - city: str('input_1')
-            - port:
-                system_property: user.sys.props.port
+            - port: "'22'"
```
Comment (meirwah) [inline cloudslang-compiler/src/test/resources/flow_implicit_alias_for_current_namespace.sl]: why not : 
- port:
    system_property: user.sys.props.port
???

**Comment by meirwah** [comment]
looks good!

**Comment by meirwah** [comment]
also please update @Bonczidai ...

File: cloudslang-compiler/src/test/resources/flow_implicit_alias_for_current_namespace.sl (hunk 4 above)
Comment (Bonczidai) [inline cloudslang-compiler/src/test/resources/flow_implicit_alias_for_current_namespace.sl]: task args don't support `system property` modifier

**Comment by Bonczidai** [comment]
updated

**Comment by meirwah** [comment]
we need to keep this in a branch for now.. until we will align all content to it...

**Comment by Bonczidai** [comment]
merged from master, handled conflicts, added error message for a test

**Comment by meirwah** [comment]
So I can merge this right @Bonczidai ? nothing is left on this PR (outside of merging the relevant PR to content too)

**Comment by Bonczidai** [comment]
correct, it can be merged (release also will be needed in content PR)

**Comment by meirwah** [comment]
:+1:

**Pull request merged**